### ISSUE TITLE
[KeyboardAvoidingView] Cannot handle several `TextInput`s at the same time? Question

### ISSUE THREAD

**Issue by Swordsman-Inaction** [facebook/react-native#8855]
react-native-cli: 0.2.0
react-native: 0.29.2

IOS, on Mac

Use `KeyboardAvoidingView` according to:
**Source:**
https://github.com/facebook/react-native/blob/0.29-stable/Libraries/Components/Keyboard/KeyboardAvoidingView.js
**Example:**
https://github.com/facebook/react-native/blob/0.29-stable/Examples/UIExplorer/KeyboardAvoidingViewExample.js

**Problem:**
I have a `View` which contains several `TextInput`s, so I added `KeyboardAvoidingView` on the root of the view, like this: 

```
<KeyboardAvoidingView behavior="height" style={styles.container}>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <Text style={styles.welcome}>
          Welcome to React Native!
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="O"
        />
        <Text style={styles.instructions}>
          To get started, edit index.ios.js
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <Text style={styles.instructions}>
          Press Cmd+R to reload,{'\n'}
          Cmd+D or shake for dev menu
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="LOL"
        />
      </KeyboardAvoidingView>
```

I'v tried all three behaviors: `position`, `height` and `padding`, they all act strangely:
**position**

![position](https://cloud.githubusercontent.com/assets/12252463/16906168/36b650ee-4ce5-11e6-966f-e7065ef8fb41.gif)

**height**

![height](https://cloud.githubusercontent.com/assets/12252463/16906171/45eb260c-4ce5-11e6-8b09-1a68c4ea88ac.gif)

**padding**

![padding](https://cloud.githubusercontent.com/assets/12252463/16906177/4b0cc0f0-4ce5-11e6-97ff-e73241ba8700.gif)

Then I tried to just wrap the last `TextInput` with `KeyboardAvoidingView`, like this:

```
      <View style={styles.container}>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <Text style={styles.welcome}>
          Welcome to React Native!
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="O"
        />
        <Text style={styles.instructions}>
          To get started, edit index.ios.js
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <Text style={styles.instructions}>
          Press Cmd+R to reload,{'\n'}
          Cmd+D or shake for dev menu
        </Text>
        <KeyboardAvoidingView behavior="position" style={{alignSelf: 'stretch'}}>
          <TextInput
            style={styles.textInput}
            placeholder="lol"
            defaultValue="LOL"
          />
        </KeyboardAvoidingView>
      </View>
```

when I choose `position` for `behavior`, it's like this: 

![part_position](https://cloud.githubusercontent.com/assets/12252463/16906213/2b4fd6ca-4ce6-11e6-8dd1-d349e7001129.gif)

Seems like every `TextInput` can trigger the `KeyboardAvoidingView`'s behavior.

**Another problem:** 
When `KeyboardAvoidingView` is wrapped with a `position: 'absolute'` `View`, it doesn't work at all, like this:

```
      <View style={styles.container}>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <Text style={styles.welcome}>
          Welcome to React Native!
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="O"
        />
        <Text style={styles.instructions}>
          To get started, edit index.ios.js
        </Text>
        <TextInput
          style={styles.textInput}
          placeholder="lol"
          defaultValue="L"
        />
        <View style={{position: 'absolute', top: 470, right: 0, left: 0, bottom: 0}}>
          <Text style={styles.instructions}>
            Press Cmd+R to reload,{'\n'}
            Cmd+D or shake for dev menu
          </Text>
          <KeyboardAvoidingView behavior="position" style={{alignSelf: 'stretch'}} >
            <TextInput
              style={styles.textInput}
              placeholder="lol"
              defaultValue="LOL"
            />
          </KeyboardAvoidingView>
        </View>
      </View>
```

Behavior:

![inside_position](https://cloud.githubusercontent.com/assets/12252463/16906332/a51309d0-4ce8-11e6-8c60-9e4bcbc8e927.gif)

**Question:**
Did I use `KeyboardAvoidingView` in a wrong way? Or it can not handle multiple `TextInput`s

**Comment by Swordsman-Inaction** [comment]
`https://github.com/APSL/react-native-keyboard-aware-scroll-view` is a great one.

**Comment by sibelius** [comment]
@nicklockwood `KeyboardAvoidindView` can handle this scenario?

**Comment by inetfuture** [comment]
And why it's not documented? Is it because it's not production ready?

**Comment by lprhodes** [comment]
I'm seeing the same thing - It'd be great to get some comment on whether we can use KeyboardAvoidindView and whether it supports multiple text fields

**Comment by qualifyapp** [comment]
This issue should not have been closed with a link to a third party solution.

**Comment by Swordsman-Inaction** [comment]
Reopen it since it still exists.

**Comment by jhack32** [comment]
Try adjusting the style
For example, <KeyboardAvoidingView behavior="position" style={{alignItems: 'center',}}>

**Comment by jrwm** [comment]
@Swordsman-Inaction I am encoutering the same situation as you. This problem persists in version 0.39.2. APSL/react-native-keyboard-aware-scroll-view works fine though. Thanks you for sharing it ;)

**Comment by keeleycarrigan** [comment]
Yea it seems like `KeyboardAvoidingView` isn't working for me at all, and I think @Swordsman-Inaction pointed out why. The views they are in are absolutely positioned. This is (I believe) impossible to avoid when using `NavigationExperimental` because all the scenes are absolutely positioned to do the animations properly.

**Comment by otoinsa** [comment]
This component does not work in any expected way, it surely does some weird stuff;

It would be good to update documentation with multiple inputs/scroll view, if we are not using it right. However I tried (seemingly) all the possible scenarios, also setting contentContainerStyle; just got really tired of it after 3 hrs of no acceptable result... 

Maybe you should implement keyboard-aware-scroll approach :D

**Comment by anahitsk123** [comment]
Is issue solved yet ? I still see it.
The keyboard is covering the TextInput.

**Comment by buskerone** [comment]
cmon guys I think this is a very important feature in an mobile app! ESSENTIAL!

**Comment by binzailani3136** [comment]
Who solved this important problem?
https://github.com/APSL/react-native-keyboard-aware-scroll-view is working well on iOS as expected but it's not working on Android.
I've tried everything what I can do including `KeyboardAvoidingView` and other libraries.
But none of them working well on Android.
Is there any library which supports Android very well?

**Comment by drblmb** [comment]
KeyboardAvoidingView is very hard to get working.  On Android, you don't need it at all thanks to the new android:windowSoftInputMode="adjustResize" setting in the AndroidManifest.xml file.  For all of my use cases, this setting works beautifully on Android.

When you have absolute or fixed-size views, KeyboardAvoidingView is simply not working.

**Comment by jdgarzon64** [comment]
https://github.com/drblmb thanks, you solve this problllem on my android device :D

**Comment by bethuel11** [comment]
@drblmb  it works perfectly.. thanks

**Comment by alvaromb** [comment]
I'm the author of `react-native-keyboard-aware-scroll-view`. I would be happy to:

- Integrate my solution into RN core.
- Help to improve `KeyboardAvoidingView` with a similar approach than the one I've used.
- Help to improve `KeyboardAvoidingView`.
- Keep maintaining my component as a separated repo.

And I still don't think that linking to a third party solution should close the issue. Keyboard handling should live into the core, IMHO.

**Comment by douglasjunior** [comment]
I got a good result using `android:windowSoftInputMode="adjustResize"` for Android and [IQKeyboardManager](https://github.com/hackiftekhar/IQKeyboardManager) for iOS.

**Comment by douglasjunior** [comment]
I got a good result using `android:windowSoftInputMode="adjustResize"` for Android and [IQKeyboardManager](https://github.com/hackiftekhar/IQKeyboardManager) for iOS.

I also created a [library](https://github.com/douglasjunior/react-native-keyboard-manager) for ease of use. 

Just install, no extra code needed:
```bash
npm install --save react-native-keyboard-manager
```

**Comment by stoffern** [comment]
Ios!! Not android

**Comment by douglasjunior** [comment]
Yes, my library os for Ios.

**Comment by shergin** [comment]
`KeyboardAvoidingView` does NOT handle `TextInput`s at all (separately or not).
The whole purpose of KeyboardAvoidingView is reacting on software keyboard appearance events and adjusting **own** layout accordingly. It can change own `position`, `size`, and `padding` (the most common use case).

`KeyboardAvoidingView` does NOT change `ScrollView` padding (aka insets).
`KeyboardAvoidingView` does NOT scroll `ScrollView`s to focused `TextInput`s.

So, usually `KeyboardAvoidingView` should be used as top-level view container.

**Comment by baijunjie** [comment]
https://github.com/baijunjie/react-native-input-scroll-view

**Comment by Jp3rd** [comment]
@drblmb Hi, is there any example code for you implementation of the android side? Cos I have  `android:windowSoftInputMode="adjustResize"` in. But it still doesn't work. Or am I doing it wrong?


`<KeyboardAvoidingView 
      behavior="padding" 
      style={Styles.mainContainer}>

        <View style={Styles.container}>
          <View style={Styles.row}>
            <View style={Styles.logoContainer} >
              <View style={Styles.logo} />
            </View>
            <View style={Styles.inputsContainer} >
              <View
              style={Styles.inputsWhiteBox} >

                <View style={Styles.usernameContainer} >
                  <TextInput
                    ref='username'
                    style={textInputStyle}
                    value={username}
                    editable={editable}
                    keyboardType='default'
                    returnKeyType='next'
                    autoCapitalize='none'
                    autoCorrect={false}
                    onChangeText={this.handleChangeUsername}
                    underlineColorAndroid='transparent'
                    onSubmitEditing={() => this.refs.password.focus()}
                    placeholder='Username' />
                </View>

                <View style={Styles.lineContainer} >
                  <Text style={Styles.line} />
                </View>

                <View style={Styles.usernameContainer} >
                  <TextInput
                    ref='password'
                    style={textInputStyle}
                    value={password}
                    editable={editable}
                    keyboardType='default'
                    returnKeyType='go'
                    autoCapitalize='none'
                    autoCorrect={false}
                    secureTextEntry
                    onChangeText={this.handleChangePassword}
                    underlineColorAndroid='transparent'
                    onSubmitEditing={this.handlePressLogin}
                    placeholder='Password' />
                </View>
              </View>
            </View >
            
              <View style={Styles.errorMessageContainer}>
                <ErrorMessage errorMessage={this.props.errorMessage} />
              </View>
            
            <View style={Styles.buttonsContainer} >
              <View style={Styles.flexBox}>
                <TouchableOpacity style={Styles.loginButtonWrapper} onPress={this.handlePressLogin}>
                  <View style={Styles.blueBox}>
                    <Text style={Styles.loginText}>Login</Text>
                  </View>
                </TouchableOpacity>
              </View>

              <View style={Styles.flexBox}>
                <TouchableOpacity style={Styles.cancelButtonWrapper} onPress={() => this.props.back()}>
                  <View style={Styles.blueBox}>
                    <Text style={Styles.loginText}>Cancel</Text>
                  </View>
                </TouchableOpacity>
              </View>
            </View>
            
            <View style={Styles.loginLaterTextContainer} >
              <Text style={Styles.footerText}>Want to browse first? <Text style={Styles.loginLaterText}
                onPress={this.props.onPressLoginLater} >
                Log in Later
              </Text>
              </Text>
            </View>
          </View>
        </View>
      </KeyboardAvoidingView>
`

**Comment by johnfdhartman** [comment]
This is still a problem.

**Comment by jomla97** [comment]
+1. This is still very much an issue. Not to even speak of the documentation, which is lacking in a lot of ways.

**Comment by drblmb** [comment]
@Jp3rd - sorry for the late reply.  You don't use <KeyboardAvoidingView /> at all on Android.  Just use the adjustResize setting.

**Comment by CorinneKelly** [comment]
This is definitely still an issue. If there is a solution it has not been made clear here or in the documentation. I think the most common mobile behavior that developers AND users are looking for is to be able to have a user click on a textInput and have the screen slide up so that the active input box is on top of the keyboard. This is easy to accomplish when theres only one textinput on the screen but often there are multiple textInputs on the screen and all three "behaviors" of KeyboardAvoidingView are not useful at all. Please reopen this or at least share a solution if there is one.

**Comment by douglasjunior** [comment]
The problem exists in iOS because the OS itself does not offer a solution, even when we develop native solutions we suffer with it.

When I worked with native iOS apps, I used the [`IQKeyboardManager`](https://github.com/hackiftekhar/IQKeyboardManager) for a long time. So, after migrate to RN, I created the [`react-native-keyboard-manager`](https://github.com/douglasjunior/react-native-keyboard-manager) and it has worked very well.

I think the RN will not create a solution for this now, it's not a trivial thing to do.
